src/annotator/highlight-clusters.tsx

```diff
@@ -113,6 +134,83 @@ export class HighlightClusterController implements Destroyable {
     this._activate(this._isActive());
   }
 
+  /**
+   * Update nesting data- attributes on <hypothesis-highlight> and associated
+   * SVG highlight rects.
+   */
+  _updateClusterData() {
+    if (!this._isActive()) {
+      return;
+    }
+
+    /**
+     * Recursively walk a tree of <hypothesis-highlight> elements, adding
+     * nesting data attributes to both <hypothesis-highlight> and their
+     * associated SVG highlight (rect) elements. This allows styling based on
+     * nesting relationships and depth.
+     *
+     * @param highlightEls - A collection of sibling <hypothesis-highlight>
+     * elements
+     * @param parentCluster - The cluster value of the parent highlight to
+     * `highlightEls`, if any
+     * @param nestingLevel - The nesting "level", relative to the outermost
+     * <hypothesis-highlight> element (0-based)
+     * @param parentClusterLevel - The parent's nesting depth, per its cluster
+     * value (`parentCluster`). i.e. How many levels since the cluster value
+     * changed? This allows for nested styling of highlights of the same cluster
+     * value.
+     */
+    const _updateHighlightClusters = (
+      highlightEls: HighlightElement[],
+      parentCluster = '',
+      nestingLevel = 0,
+      parentClusterLevel = 0
+    ) => {
+      Array.from(highlightEls).forEach(hEl => {
+        const elCluster =
+          ['user-annotations', 'user-highlights', 'other-content'].find(
+            candidate => hEl.classList.contains(candidate)
+          ) ?? 'other-content';
+
+        const elClusterLevel =
+          parentCluster && elCluster === parentCluster
+            ? parentClusterLevel + 1
+            : 0;
+
+        hEl.setAttribute('data-nesting-level', `${nestingLevel}`);
```

Comment:
The `data-nesting-level` attribute is currently written but not read in CSS/JS. Do you have a plan to use it in future?

Reply:
There are a few reasons I added this:

1) Data symmetry between the `<hypothesis-highlight>` and associated SVG `rect` elements;
2) A hook for the future — possibly even the immediate future as I look at possible "blending" styling;
3) We could opt to _only_ style nested highlights with `data-` attributes (i.e. remove the hierarchical selectors from the CSS). This would have the advantage of consolidation and equivalence of experience across document types and simplification of CSS, but would lose the immediacy of nested highlight styling currently enjoyed in HTML documents.

I think the way to make some decisions here is to feel out whether data attributes are useful on `hypothesis-highlight` elements over the next week or two, as we build this out more.